Issue 275: Alternative approach using a listener

WilliamParker opened the pull request:
I gave the previous PR https://github.com/cerner/clara-rules/pull/425 some thought and on reflection, I think this can be done in a simpler way that doesn't impact as much of the core code with a listener.  The actual behaviour upon reaching the limit could then be easily user-defined, probably with some default options supplied.

Thoughts?  There is definitely some stuff to clean up here before merging but I wanted to get thoughts on the general approach before getting too far into this. The tests are fundamentally the same as in https://github.com/cerner/clara-rules/pull/425 with some minor adjustments to invoke the listener; in a final PR I'd change them further to not solely rely on exceptions being thrown, which would allow them to be cross-platform.

@EthanEChristian @mrrodriguez

mrrodriguez commented on src/test/clojure/clara/test_infinite_loops.clj:
```diff
@@ -0,0 +1,189 @@
+;;           (is (= (-> loop-data first  :listeners count) 1)
+;;               "There should only be one listener.")
+;;           (is (-> loop-data first  :listeners ^PersistentTracingListener first .-trace not-empty)
+;;               "There should be tracing data available when a traced session throws an exception on an infinite loop."))))))
```
This is still an issue with the listener approach?
I'm not sure there necessarily should/can be a trace available when arbitrary exceptions are thrown during rule evaluation.

If we want that, I think that's a broader issue than just specifically having a trace when this max cycle logic throws.

WilliamParker replied:
If there is an infinite loop I think it is useful to able to use a trace to show which rules are looping.  My approach in https://github.com/cerner/clara-rules/pull/458 allows the trace to be added.

EthanEChristian commented on src/main/clojure/clara/rules/listener.cljc:
```diff
@@ -23,6 +23,7 @@
   (remove-activations! [listener node activations])
   (fire-activation! [listener activation resulting-operations])
   (fire-rules! [listener node])
+  (activation-group-transition! [listener original-group new-group])
```
I agree with the premise here, we've had breaking changes to listeners in the past and while not a big deal it still impacts consumers.
> This isn't an issue to be solved in this particular work.
It might be worth it to log an issue to look into ways that we can change listeners in the future to be less impactful. For now i think i am ok with adding a new method to this protocol, and announcing it as a breaking change, like we did in the past.

WilliamParker replied:
I've logged https://github.com/cerner/clara-rules/issues/457 to look into a better approach to this with a thought on how to do it.  For now I think we can do as Ethan suggested above.

EthanEChristian commented on src/main/clojure/clara/tools/loop_detector.cljc:
```diff
@@ -0,0 +1,51 @@
+  (:require [clara.rules.listener :as l]
+            [clara.rules.engine :as eng]))
+
+(deftype LoopDetectorListener [cycles-count max-cycles fn-on-limit]
```
to be more explicit it could also be said that `on-limit-exceeded-fn` would be more clear.

EthanEChristian commented on src/main/clojure/clara/tools/loop_detector.cljc:
```diff
@@ -0,0 +1,51 @@
+  (:require [clara.rules.listener :as l]
+            [clara.rules.engine :as eng]))
+
+(deftype LoopDetectorListener [cycles-count max-cycles fn-on-limit]
```
`MaxCycleListener` seems a bit vague from a new user perspective, `CyclicalRuleListener` perhaps...

WilliamParker replied:
+1 I like "CyclicalRuleListener", will change.

WilliamParker commented on src/main/clojure/clara/tools/loop_detector.cljc:
```diff
@@ -0,0 +1,51 @@
+  (to-transient [listener]
+    listener))
+
+(defn with-loop-detection [session fn-on-limit max-cycles]
```
Agreed, I have changed this in https://github.com/cerner/clara-rules/pull/458

WilliamParker commented on src/main/clojure/clara/tools/testing_utils.cljc:
```diff
@@ -203,3 +203,19 @@
+     "Given a Throwable, return in order the ExceptionInfo data maps for all items in the
+      chain that implement IExceptionInfo and have nonempty data maps."
+     [t]
+     (let [throwables  ((fn append-self
```
+1, changed in https://github.com/cerner/clara-rules/pull/458

WilliamParker commented on src/main/clojure/clara/tools/testing_utils.cljc:
```diff
@@ -203,3 +203,19 @@
                      res#)))))))
+
+#?(:clj
+   (defn ex-data-maps
```
I've used a slightly modified version of this docstring in https://github.com/cerner/clara-rules/pull/458

Note that I think this can return empty ex-data maps; I have changed the docstring you proposed above accordingly.

WilliamParker commented:
I had some difficulties with Git merge conflicts when trying to add changes on this PR, so I have created a new PR branched off master at https://github.com/cerner/clara-rules/pull/458 that addresses the comments in this PR.

Closing in favour of that PR.
